src/main/java/us/abstracta/wiresham/VirtualTcpBase.java

```diff
@@ -0,0 +1,118 @@
+package us.abstracta.wiresham;
+
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import javax.net.ssl.SSLContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class VirtualTcpBase<T> {
+
+  public static final int DEFAULT_READ_BUFFER_SIZE = 2048;
+  public static final int DYNAMIC_PORT = 0;
+
+  private static final Logger LOG = LoggerFactory.getLogger(VirtualTcpBase.class);
+
+  protected final Set<ConnectionFlowStepsDriver> connectionDrivers = new HashSet<>();
+  protected final Map<List<PacketStep>, T> packetsByPortMap = new LinkedHashMap<>();
+  protected SSLContext sslContext;
+  protected int readBufferSize = DEFAULT_READ_BUFFER_SIZE;
+  protected int port = DYNAMIC_PORT;
+  protected ExecutorService virtualExecutorService;
+  private boolean sslEnabled;
+
+  public void configureFlow(Flow flow)
+      throws IOException, NoSuchAlgorithmException, KeyManagementException {
+    verifyMaxLengthPacket(flow);
+    buildDissectedPacketsByPort(flow);
+  }
+
+  private void verifyMaxLengthPacket(Flow flow) {
+    Optional<PacketStep> bigPacketStep = flow.getSteps().stream()
+        .filter(s -> s instanceof ReceivePacketStep && s.data.getBytes().length > readBufferSize)
+        .findAny();
+    if (bigPacketStep.isPresent()) {
+      throw new IllegalArgumentException(String.format(
+          "Read buffer size of %d bytes is not enough for receiving expected packet from client "
+              + "with %s", readBufferSize, bigPacketStep.get().data));
+    }
+  }
+
+  protected void buildDissectedPacketsByPort(Flow flow)
+      throws IOException, NoSuchAlgorithmException, KeyManagementException {
+    List<PacketStep> currentFlowPort = new ArrayList<>();
```

Comment:
You keep creating "sub flows" by ports? why? is it simpler than having a flow with each step consuming/producing to a particular socket, and some other entity (the virtualTCPService or similar) creating or assigning sockets to flows when connections reach the set of server sockets?

Reply:
I think I may have missed that VirtualTcpClient creates sockets and not serverSockets, but in such scenario you just give the packet a socket provider, in one case creating the socket, and in the other waiting for a socket generated after a serverSocket connection

Reply:
I still need to review the full extent of current implementation...